pipx/pipxrc.py

```diff
@@ -0,0 +1,196 @@
+
+    def get_package_or_url(self, default: str) -> str:
+        return self._val_or_default(self.pipxrc_info.package_or_url, default)
+
+    def get_install_pip_args(self, default: List[str]) -> List[str]:
+        return self._val_or_default(self.pipxrc_info.install.pip_args, default)
+
+    def get_install_venv_args(self, default: List[str]) -> List[str]:
+        return self._val_or_default(self.pipxrc_info.install.venv_args, default)
+
+    def get_install_include_dependencies(self, default: bool) -> bool:
+        return self._val_or_default(
+            self.pipxrc_info.install.include_dependencies, default
+        )
+
+    def get_venv_metadata(self, default: PipxVenvMetadata) -> PipxVenvMetadata:
+        return self._val_or_default(self.pipxrc_info.venv_metadata, default)
+
+    def get_injected_packages(
+        self, default: List[Dict[str, Any]]
+    ) -> List[Dict[str, Any]]:
+        if self.pipxrc_info.injected_packages is not None:
+            injected_packages = []
+            for package in self.pipxrc_info.injected_packages:
+                package_info = {"package": package}
+                package_info.update(
+                    self.pipxrc_info.injected_packages[package]._asdict()
+                )
+                injected_packages.append(package_info)
+            return injected_packages
+        else:
+            return default
+
+    def set_package_or_url(self, package_or_url: str) -> None:
+        # TODO 20190923: if package_or_url is a local path, we need to make it
+        #   an absolute path
+        self.pipxrc_info.package_or_url = package_or_url
+
+    def set_venv_metadata(self, venv_metadata: PipxVenvMetadata) -> None:
+        self.pipxrc_info.venv_metadata = venv_metadata
+
+    def set_install_options(
+        self, pip_args: List[str], venv_args: List[str], include_dependencies: bool
+    ) -> None:
+        self.pipxrc_info.install = InstallOptions(
+            pip_args=pip_args,
+            venv_args=venv_args,
+            include_dependencies=include_dependencies,
+        )
+
+    def add_injected_package(
+        self,
+        package: str,
+        pip_args: List[str],
+        verbose: bool,
+        include_apps: bool,
+        include_dependencies: bool,
+        force: bool,
+    ) -> None:
+        if self.pipxrc_info.injected_packages is None:
+            self.pipxrc_info.injected_packages = {}
+
+        self.pipxrc_info.injected_packages[package] = InjectedPackage(
+            pip_args=pip_args,
+            verbose=verbose,
+            include_apps=include_apps,
+            include_dependencies=include_dependencies,
+            force=force,
+        )
+
+    def write(self) -> None:
+        # If writing out, make sure injected_packages is not None, so next
+        #   successful read of pipxrc does not use default in
+        #   get_injected_packages()
+        if self.pipxrc_info.injected_packages is None:
+            self.pipxrc_info.injected_packages = {}
```

Comment:
Maybe just default to `{}` so we don't have to be careful everywhere about this?

Reply:
This was a thorny topic that I was wrestling with.

The reason for all the `None`s was to have a way to say "This information is missing," as opposed to "There were never any injected packages," etc.  I was trying to account for the possibility that the calling code had never read in the original pipxrc, or for a new pipxrc, had never filled in particular information in the first place.

I suppose in this case if one really wanted to be strict, the code could raise an exception on `write()` if there were any `None`'s left, as a means of self-validation, instead of changing it to a `{}` as the code above does.

I'm curious what you think of the above paragraphs.

Reply:
But maybe injected_packages is a special case and could just default to {}.  There's currently no way to set it to {} through the class methods.

Reply:
> The reason for all the Nones was to have a way to say "This information is missing," as opposed to "There were never any injected packages," etc. I was trying to account for the possibility that the calling code had never read in the original pipxrc, or for a new pipxrc, had never filled in particular information in the first place.

Gotcha, if that's the case then it's fine. It was just something that caught my eye.